### PR TITLE
Refactor CommittableKafkaWriter to use non-blocking request handling

### PR DESCRIPTION
**Pull request by kabochya**
tl; dr: Replaced `clusterThreadPool` usage by using CompletableFutures to implement non-blocking kafka request handling

In this diff, we remove the usage of clusterThreadPool in `CommitableKafkaWriter`. Previously, the `clusterThreadPool` was allocated per cluster and used for tracking metrics of the `Future<RecordMetadata>`. This caused a lot of threads to be used if Singer is publishing to a lot of clusters. Since these threads are used only for tracking, we found out that they could be eliminated and instead refactored into `CompletableFuture`s that can be executed asynchronously with the shared common `ForkJoinPool` as long as they remain non-blocking. 
We also leverage the callback in `send` to achieve this:
```
    CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
    committableProducer.send(keyedMessage, (recordMetadata, exception) -> {
      if (exception != null) {
        future.completeExceptionally(exception);
      } else {
        future.complete(recordMetadata);
      }
    });
```
 After this diff, it is expected to drop the number of threads created by Singer and significantly reduce the native memory allocated for multi-cluster-publishing Singer instances.

### CODE REVIEW

File: singer/src/main/java/com/pinterest/singer/writer/kafka/CommittableKafkaWriter.java [hunk 1]
```diff
@@ -138,20 +148,20 @@ public void startCommit() throws LogStreamWriterException {
       }
     }
 
-    commitableBuckets = new HashMap<>();
-    committableMapOfTrackedMessageMaps = new HashMap<>();
-    committableMapOfInvalidMessageMaps = new HashMap<>();
-    committableMapOfOriginalIndexWithinBucket = new HashMap<>();
-
+    committableBuckets.clear();
```
Comment (ambud) [inline singer/src/main/java/com/pinterest/singer/writer/kafka/CommittableKafkaWriter.java:151]: In my past experience it's better to create a new data structure from GC cleanliness stand point.

File: singer/src/main/java/com/pinterest/singer/writer/kafka/CommittableKafkaWriter.java [hunk 2]
```diff
@@ -138,20 +148,20 @@ public void startCommit() throws LogStreamWriterException {
...
-    for (int i = 0; i < committableValidPartitions.size(); i++) {
+    for (PartitionInfo partitionInfo : committableValidPartitions) {
       // for each partitionId, there is a corresponding bucket in buckets and a
       // corresponding headersMap in mapOfHeadersMaps.
-      PartitionInfo partitionInfo = committableValidPartitions.get(i);
       int partitionId = partitionInfo.partition();
-      commitableBuckets.put(partitionId, new KafkaWritingTaskFuture(partitionInfo));
-      committableMapOfTrackedMessageMaps.put(partitionId, new HashMap<Integer, LoggingAuditHeaders>());
-      committableMapOfInvalidMessageMaps.put(partitionId, new HashMap<Integer, LoggingAuditHeaders>());
+      committableBuckets.put(partitionId, new KafkaWritingTaskFuture(partitionInfo));
```
Comment (ambud) [inline singer/src/main/java/com/pinterest/singer/writer/kafka/CommittableKafkaWriter.java:160]: Why was there an indentation update?

**Review by ambud — COMMENTED** [review]
Added initial comments

**Comment by ambud** [comment]
Could you please add a detail description of what this PR is about and how it improves singer?

File: singer/src/main/java/com/pinterest/singer/writer/kafka/CommittableKafkaWriter.java (hunk 2 above)
Comment (kabochya) [reply singer/src/main/java/com/pinterest/singer/writer/kafka/CommittableKafkaWriter.java:160]: It was actually just fixing the `commitableBuckets` typo to `committableBuckets`

**Comment by ambud** [comment]
Could you also include a version update?

**Review by ambud — APPROVED** [review]
